File: leetcode/src/01220_count_vowels_permutation/solution.go
package leetcode

func countVowelPermutation(n int) int {
	const m = 1_000_000_007
	a, e, i, o, u := 0, 1, 2, 3, 4
	var dp [5][]int
	for i := 0; i < 5; i++ {
		dp[i] = make([]int, n+1)
		dp[i][1] = 1
	}
	for j := 2; j <= n; j++ {
		dp[a][j] = dp[e][j-1]
		dp[e][j] = (dp[a][j-1] + dp[i][j-1]) % m
		dp[i][j] = (dp[a][j-1] + dp[e][j-1] + dp[o][j-1] + dp[u][j-1]) % m
		dp[o][j] = (dp[i][j-1] + dp[u][j-1]) % m
		dp[u][j] = dp[a][j-1]
	}
	res := 0
	for _, ele := range dp {
		res = (res + ele[n]) % m
	}
	return res
}

File: leetcode/src/01220_count_vowels_permutation/solution.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn count_vowel_permutation(n: i32) -> i32 {
        let rules = HashMap::from([
            (' ', vec!['a', 'e', 'i', 'o', 'u']),
            ('a', vec!['e']),
            ('e', vec!['a', 'i']),
            ('i', vec!['a', 'e', 'o', 'u']),
            ('o', vec!['i', 'u']),
            ('u', vec!['a']),
        ]);
        let mut dp = vec![HashMap::<char, i32>::new(); n as usize + 1];
        Solution::solve(n, ' ', &rules, &mut dp)
    }

    fn solve(
        n: i32,
        prev: char,
        rules: &HashMap<char, Vec<char>>,
        dp: &mut Vec<HashMap<char, i32>>,
    ) -> i32 {
        if n == 0 {
            return 1;
        }
        if let Some(&val) = dp[n as usize].get(&prev) {
            return val;
        }
        let mut res = 0;
        for &cur in rules.get(&prev).unwrap_or(&vec![]) {
            res = (res + Solution::solve(n - 1, cur, rules, dp) % 1_000_000_007) % 1_000_000_007;
        }
        dp[n as usize].insert(prev, res);
        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let n = 1;
        assert_eq!(Solution::count_vowel_permutation(n), 5);

        let n = 2;
        assert_eq!(Solution::count_vowel_permutation(n), 10);

        let n = 5;
        assert_eq!(Solution::count_vowel_permutation(n), 68);
    }
}
